### PR TITLE
 r/aws_iam_user_login_profile  - Make the PGP key in optional

### LINKED ISSUE #4564: 🚀 Feature Request: Make the PGP key in aws_iam_user_login_profile optional
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The PGP password encryption functionality forces me to interact with keys and extra encryption steps, while my state file also contains lots of other secrets which are just as sensitive without these additional steps.

In our case, we're only using this to set the initial user passwords, which are flagged for immediate change - so it's not important to us to keep them encrypted in the state file anyway. In practice we immediately decrypt the value and send it over.

This also creates some churn when different people use the same terraform workspace, as they all need to provide their own PGP key, which in turn causes each resource to be modified. It looks like the code has no `Update` handler, so I presume this means that the password remains unreadable until the original person reappears with the PGP key.

### New or Affected Resource(s)

* aws_iam_user_login_profile

### Potential Terraform Configuration

```hcl
# n/a
```

### References

* If #4412 is fixed then the PGP key changes could be problematic if it forces a new password unexpectedly.

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #4564

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSUserLoginProfile_'
--- PASS: TestAccAWSUserLoginProfile_notAKey (34.68s)
--- PASS: TestAccAWSUserLoginProfile_keybaseDoesntExist (36.00s)
--- PASS: TestAccAWSUserLoginProfile_disappears (51.75s)
--- PASS: TestAccAWSUserLoginProfile_no_pgp (56.59s)
--- PASS: TestAccAWSUserLoginProfile_keybase (56.59s)
--- PASS: TestAccAWSUserLoginProfile_PasswordLength (57.24s)
--- PASS: TestAccAWSUserLoginProfile_basic (65.99s)
...
```

Cant test this unfortunately as i don't have delete user permissions in my account :(

### CODE REVIEW

File: aws/resource_aws_iam_user_login_profile_test.go [hunk 1]
```diff
@@ -92,10 +92,40 @@ func TestAccAWSUserLoginProfile_basic(t *testing.T) {
...
+					testAccCheckAWSUserLoginProfileExists(resourceName, &conf),
+					resource.TestCheckResourceAttr(resourceName, "password_length", "20"),
+					resource.TestCheckResourceAttr(resourceName, "password_reset_required", "true"),
+				),
+			},
+			{
+				ResourceName:            resourceName,
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"password_length", "password_reset_required"},
```
Comment (ewbankkit) [inline aws/resource_aws_iam_user_login_profile_test.go]:
```console
--- FAIL: TestAccAWSUserLoginProfile_no_pgp (27.20s)
    testing.go:654: Step 1 error: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.
        
        (map[string]string) (len=2) {
         (string) (len=18) "encrypted_password": (string) "",
         (string) (len=15) "key_fingerprint": (string) ""
        }
        
        
        (map[string]string) {
        }
```

Use the same as other import tests?

```go
				ImportStateVerifyIgnore: []string{
					"encrypted_password",
					"key_fingerprint",
					"password_length",
					"password_reset_required",
					"pgp_key",
				},
```

File: aws/resource_aws_iam_user_login_profile.go [hunk 2]
```diff
@@ -152,8 +141,22 @@ func resourceAwsIamUserLoginProfileCreate(d *schema.ResourceData, meta interface
...
+		if err != nil {
+			return err
+		}
+		fingerprint, encrypted, err := encryption.EncryptValue(encryptionKey, initialPassword, "Password")
+		if err != nil {
+			return err
+		}
+
+		d.Set("key_fingerprint", fingerprint)
+		d.Set("encrypted_secret", encrypted)
```
Comment (ewbankkit) [inline aws/resource_aws_iam_user_login_profile.go]: Should be `encrypted_password`:

```console
--- FAIL: TestAccAWSUserLoginProfile_basic (16.23s)
    testing.go:654: Step 0 error: Check failed: Check 2/7 error: No password in state
```

**Comment by ewbankkit** [comment]
Acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSUserLoginProfile_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccAWSUserLoginProfile_ -timeout 120m
=== RUN   TestAccAWSUserLoginProfile_basic
=== PAUSE TestAccAWSUserLoginProfile_basic
=== RUN   TestAccAWSUserLoginProfile_no_pgp
=== PAUSE TestAccAWSUserLoginProfile_no_pgp
=== RUN   TestAccAWSUserLoginProfile_keybase
=== PAUSE TestAccAWSUserLoginProfile_keybase
=== RUN   TestAccAWSUserLoginProfile_keybaseDoesntExist
=== PAUSE TestAccAWSUserLoginProfile_keybaseDoesntExist
=== RUN   TestAccAWSUserLoginProfile_notAKey
=== PAUSE TestAccAWSUserLoginProfile_notAKey
=== RUN   TestAccAWSUserLoginProfile_PasswordLength
=== PAUSE TestAccAWSUserLoginProfile_PasswordLength
=== CONT  TestAccAWSUserLoginProfile_basic
=== CONT  TestAccAWSUserLoginProfile_notAKey
=== CONT  TestAccAWSUserLoginProfile_keybaseDoesntExist
=== CONT  TestAccAWSUserLoginProfile_keybase
=== CONT  TestAccAWSUserLoginProfile_no_pgp
=== CONT  TestAccAWSUserLoginProfile_PasswordLength
--- FAIL: TestAccAWSUserLoginProfile_basic (16.23s)
    testing.go:654: Step 0 error: Check failed: Check 2/7 error: No password in state
--- FAIL: TestAccAWSUserLoginProfile_keybase (16.40s)
    testing.go:654: Step 0 error: errors during apply:
        
        Error: Error retrieving Public Key for keybase:terraformacctest
        : parse https://keybase.io/_/api/1.0/user/lookup.json?usernames=terraformacctest
        &fields=public_keys: net/url: invalid control character in URL
        
          on /tmp/tf-test249940899/main.tf line 36:
          (source code not available)
        
        
--- PASS: TestAccAWSUserLoginProfile_keybaseDoesntExist (18.28s)
--- PASS: TestAccAWSUserLoginProfile_notAKey (18.57s)
--- FAIL: TestAccAWSUserLoginProfile_no_pgp (27.20s)
    testing.go:654: Step 1 error: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.
        
        (map[string]string) (len=2) {
         (string) (len=18) "encrypted_password": (string) "",
         (string) (len=15) "key_fingerprint": (string) ""
        }
        
        
        (map[string]string) {
        }
        
--- PASS: TestAccAWSUserLoginProfile_PasswordLength (27.41s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	27.455s
FAIL
GNUmakefile:26: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

Not sure why `TestAccAWSUserLoginProfile_keybase` is failing...

**Comment by tomelliff** [comment]
@DrFaust92 I appreciate this is still at a draft stage but it looks like it needs a bit more work to actually implement it usefully (eg still return an unencrypted password when `pgp_key` isn't set) and fix tests highlighted by @ewbankkit etc.

Are you still interested in finishing this off? I'd be tempted to do the small amount of work on top of yours to get this over the line if you're unable to commit anything more towards it as it might be useful for us.

**Comment by DrFaust92** [comment]
@tomelliff thanks for bringing this up, ill give it a go again (at least ill rebase and push). if i don't get this in the next few days feel free to take over

**Comment by DrFaust92** [comment]
Fixed all test issues:

```
--- PASS: TestAccAWSUserLoginProfile_notAKey (34.68s)
--- PASS: TestAccAWSUserLoginProfile_keybaseDoesntExist (36.00s)
--- PASS: TestAccAWSUserLoginProfile_disappears (51.75s)
--- PASS: TestAccAWSUserLoginProfile_no_pgp (56.59s)
--- PASS: TestAccAWSUserLoginProfile_keybase (56.59s)
--- PASS: TestAccAWSUserLoginProfile_PasswordLength (57.24s)
--- PASS: TestAccAWSUserLoginProfile_basic (65.99s)
```

File: aws/resource_aws_iam_user_login_profile.go [hunk 3]
```diff
@@ -123,41 +123,45 @@ func checkIAMPwdPolicy(pass []byte) bool {
...
 	if err != nil {
-		return fmt.Errorf("Error creating IAM User Login Profile for %q: %s", username, err)
+		return fmt.Errorf("Error creating IAM User Login Profile for %q: %w", username, err)
+	}
+
+	d.SetId(aws.StringValue(createResp.LoginProfile.UserName))
+
+	if v, ok := d.GetOk("pgp_key"); ok {
+		pgpKey := v.(string)
+		pgpKey = strings.TrimSuffix(pgpKey, "\n")
```
Comment (DrFaust92) [inline aws/resource_aws_iam_user_login_profile.go:151]: maybe not the best way to handle this and a dedicated test config would be better

**Comment by jsmilani** [comment]
This seems to have stalled. According to https://www.terraform.io/docs/extend/best-practices/sensitive-state.html#don-39-t-encrypt-state it is necessary to deprecate PGP encrypted state. Can we keep this moving?

**Comment by DrFaust92** [comment]
Rebased 

```
--- PASS: TestAccIAMUserLoginProfile_nogpg (46.59s)
--- PASS: TestAccIAMUserLoginProfile_passwordLength (46.70s)
--- PASS: TestAccIAMUserLoginProfile_notAKey (47.34s)
--- PASS: TestAccIAMUserLoginProfile_keybaseDoesntExist (49.97s)
--- PASS: TestAccIAMUserLoginProfile_disappears (61.82s)
--- PASS: TestAccIAMUserLoginProfile_basic (63.68s)
```

keybase test fails and im unsure how to tackle it

```
    user_login_profile_test.go:110: Step 1/2 error: Error running apply: exit status 1
        Error: error retrieving GPG Key during IAM User Login Profile (tf-acc-test-4386915330115076656) creation: Error retrieving Public Key for keybase:terraformacctest
--- FAIL: TestAccIAMUserLoginProfile_keybase (48.69s)
```

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TESTS=TestAccIAMUserLoginProfile_ PKG=iam
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/iam/... -v -count 1 -parallel 20 -run='TestAccIAMUserLoginProfile_'  -timeout 180m
=== RUN   TestAccIAMUserLoginProfile_basic
=== PAUSE TestAccIAMUserLoginProfile_basic
=== RUN   TestAccIAMUserLoginProfile_keybase
=== PAUSE TestAccIAMUserLoginProfile_keybase
=== RUN   TestAccIAMUserLoginProfile_keybaseDoesntExist
=== PAUSE TestAccIAMUserLoginProfile_keybaseDoesntExist
=== RUN   TestAccIAMUserLoginProfile_notAKey
=== PAUSE TestAccIAMUserLoginProfile_notAKey
=== RUN   TestAccIAMUserLoginProfile_passwordLength
=== PAUSE TestAccIAMUserLoginProfile_passwordLength
=== RUN   TestAccIAMUserLoginProfile_nogpg
=== PAUSE TestAccIAMUserLoginProfile_nogpg
=== RUN   TestAccIAMUserLoginProfile_disappears
=== PAUSE TestAccIAMUserLoginProfile_disappears
=== CONT  TestAccIAMUserLoginProfile_basic
=== CONT  TestAccIAMUserLoginProfile_passwordLength
=== CONT  TestAccIAMUserLoginProfile_notAKey
=== CONT  TestAccIAMUserLoginProfile_keybaseDoesntExist
=== CONT  TestAccIAMUserLoginProfile_nogpg
=== CONT  TestAccIAMUserLoginProfile_disappears
=== CONT  TestAccIAMUserLoginProfile_keybase
    user_login_profile_test.go:110: Step 1/2 error: Error running apply: exit status 1
        
        Error: error retrieving GPG Key during IAM User Login Profile (tf-acc-test-1042917404291852556) creation: Error retrieving Public Key for keybase:terraformacctest
        : parse "https://keybase.io/_/api/1.0/user/lookup.json?usernames=terraformacctest\n&fields=public_keys": net/url: invalid control character in URL
        
          with aws_iam_user_login_profile.test,
          on terraform_plugin_test.tf line 35, in resource "aws_iam_user_login_profile" "test":
          35: resource "aws_iam_user_login_profile" "test" {
        
--- PASS: TestAccIAMUserLoginProfile_nogpg (26.92s)
--- PASS: TestAccIAMUserLoginProfile_passwordLength (28.38s)
--- PASS: TestAccIAMUserLoginProfile_notAKey (33.37s)
--- PASS: TestAccIAMUserLoginProfile_keybaseDoesntExist (33.42s)
--- FAIL: TestAccIAMUserLoginProfile_keybase (33.58s)
--- PASS: TestAccIAMUserLoginProfile_disappears (39.09s)
--- PASS: TestAccIAMUserLoginProfile_basic (41.84s)
FAIL
FAIL	github.com/hashicorp/terraform-provider-aws/internal/service/iam	45.927s
FAIL
make: *** [testacc] Error 1
```

Failure is unrelated to this change.

**Comment by ewbankkit** [comment]
@DrFaust92 Thanks for the contribution :tada: :clap:.

**Pull request merged**